### PR TITLE
Disable CodeStyle analyzers by default in all explicit builds of Roslyn.sln (i.e. csc/vbc invocations both inside Visual Studio and from command line prompt)

### PR DESCRIPTION
**Pull request by mavasani**
Workaround for https://github.com/dotnet/roslyn/issues/42166

This change brings down the local build time for Roslyn.sln on my machine from around `8 mins` to `5 mins`. I am going to file separate bugs with performance traces for slow IDE analyzers in CodeStyle NuGet package.

### CODE REVIEW

File: eng/targets/Imports.targets [hunk 1]
```diff
@@ -30,6 +30,12 @@
...
...IDE0026;IDE0027;IDE0028;IDE0028WithoutSuggestion;IDE0029;IDE0029WithoutSuggestion;IDE0030;IDE0030WithoutSuggestion;IDE0031;IDE0031WithoutSuggestion;IDE0032;IDE0032WithoutSuggestion;IDE0033;IDE0033WithoutSuggestion;IDE0034;IDE0034WithoutSuggestion;IDE0035;IDE0035WithoutSuggestion;IDE0036;IDE0036WithoutSuggestion;IDE0037;IDE0037WithoutSuggestion;IDE0040;IDE0040WithoutSuggestion;IDE0041;IDE0041WithoutSuggestion;IDE0042;IDE0042WithoutSuggestion;IDE0043;IDE0044;IDE0044WithoutSuggestion;IDE0045;IDE0045WithoutSuggestion;IDE0046;IDE0046WithoutSuggestion;IDE0047;IDE0048;IDE0048WithoutSuggestion;IDE0050;IDE0050WithoutSuggestion;IDE0051;IDE0052;IDE0054;IDE0054WithoutSuggestion;IDE0055;IDE0056;IDE0056WithoutSuggestion;IDE0057;IDE0057WithoutSuggestion;IDE0058;IDE0059;IDE0060;IDE0061;IDE0062;IDE0062WithoutSuggestion;IDE0063;IDE0063WithoutSuggestion;IDE0064;IDE0065;IDE0066;IDE0066WithoutSuggestion;IDE0070;IDE0070WithoutSuggestion;IDE0071;IDE0071WithoutSuggestion;IDE0072;IDE0072WithoutSuggestion;IDE0073;IDE0073WithoutSuggestion;IDE0074;IDE0074WithoutSuggestion;IDE0075;IDE0075WithoutSuggestion;IDE1005;IDE1005WithoutSuggestion;IDE1006;IDE1006WithoutSuggestion;RemoveUnnecessaryImportsFixable</NoWarn>
```
Comment (mavasani) [inline eng/targets/Imports.targets:37]: NOTE: This was auto-generated by using https://github.com/dotnet/roslyn/pull/42502, so includes every single diagnostic ID reported by the CodeStyle NuGet package.

Comment (sharwell) [reply eng/targets/Imports.targets:37]: > IDE0005_gen

This is not a configurable diagnostic. Non-configurable diagnostics should ideally be removed from this file, but if you can't remove them at least put them on a separate line.

Comment (mavasani) [reply eng/targets/Imports.targets:37]: I have put a blanket list of _all_ diagnostics reported by CodeStyle package here, and this was auto-generated. I think it is better to retain the entire list in one place. Hopefully we can soon revert this PR once #42166 has been implemented.

Comment (sharwell) [reply eng/targets/Imports.targets:37]: > this was auto-generated

Can we fix this to not include diagnostics which cannot be configured with a `<NoWarn>` property?

Comment (mavasani) [reply eng/targets/Imports.targets]: Done, thanks!

**Comment by jaredpar** [comment]
What is the motivation for doing this? Today the code style analyzers only run as part of our build correctness leg. They are disabled for all other legs. The motivation behind that split is exactly what you're laying out here: 

- Keep non-style legs fast as possible
- Have a single style / build correctness leg that does all of this heavy build enforcemen

**Comment by jaredpar** [comment]
Basically if we do this how will code style get verified as a part of our CI process? 

@sharwell

**Comment by mavasani** [comment]
> What is the motivation for doing this? 

Faster local builds for each developer.

**Comment by jaredpar** [comment]
The PR says 

> Disable CodeStyle analyzers by default **in CI** for Roslyn.sln

CI refers to continuous integration aka AzDO. This does not generally refer to developer inner loop.s

**Comment by jaredpar** [comment]
The description of the PR, the implementation and the discussion here all disagree on what this PR is doing.

**Comment by mavasani** [comment]
> CI refers to continuous integration aka AzDO. This does not generally refer to developer inner loop.s

Fixed now, I meant to state disable by default for _command line builds_

File: eng/targets/Imports.targets [hunk 2]
```diff
@@ -30,6 +30,12 @@
   <PropertyGroup Condition="'$(DisableNullableWarnings)' == 'true'">
     <NoWarn>$(NoWarn);Nullable</NoWarn>
   </PropertyGroup>
+  
+  <!-- Workaround for https://github.com/dotnet/roslyn/issues/42166 -->
+  <!-- PERF: This property group prevents the analyzers from CodeStyle NuGet package from being executed by default in CI -->
+  <PropertyGroup Condition="'$(DesignTimeBuild)' != 'true'">
```
Comment (jaredpar) [inline eng/targets/Imports.targets:36]: If we want this to be enabled for continuous integration only then we should use `'$(ContinuousIntegration)' != 'true'`. That is the standard way of conditioning data on CI.

Comment (mavasani) [reply eng/targets/Imports.targets:36]: That would also disable all these analyzers for design time builds, we do not want that.

**Comment by jaredpar** [comment]
> Fixed now, I meant to state disable by default for command line builds

Why command line builds here and not Visual Studio? The biggest source of complaints I've seen about code style build time is in Visual Studio, not the command line. In the command line if you don't care about code style then just use `-skipAnaylzers`. It's been there for a long time specifically for this case.

File: eng/targets/Imports.targets (hunk 2 above)
Comment (jaredpar) [reply eng/targets/Imports.targets:36]: Who is "we" here? This is definitely what a good portion of the compiler team wants for exactly what you laid out in this email. I want my inner loop fast including building from Visual Studio.

**Comment by mavasani** [comment]
> Why command line builds here and not Visual Studio? The biggest source of complaints I've seen about code style build time is in Visual Studio, not the command line. In the command line if you don't care about code style then just use -skipAnaylzers. It's been there for a long time specifically for this case.

There are about 70+ code style analyzers in this NuGet package, all of them reporting hidden diagnostics by default. We have only enforced 2 analyzers from this NuGet package as warnings in build (formatting and file header analyzer), but due to https://github.com/dotnet/roslyn/issues/42166, all of these 70+ analyzers are running on every command line build of Roslyn on every developer's machine, adding to build time overhead. Except for the explicitly enforced analyzers (current the two mentioned above), execution of rest of code style analyzers is just an unnecessary build time overhead - hidden diagnostics are never reported on command line build. This change aims at removing that overhead from command line builds.

File: eng/targets/Imports.targets (hunk 2 above)
Comment (mavasani) [reply eng/targets/Imports.targets:36]: design time build == live analysis during typing. If we make your suggested change, we will not execute these analyzers even during typing and see no suggestions.

**Comment by jaredpar** [comment]
@mavasani 

>  but due to #42166, all of these 70+ analyzers are running on every command line build of Roslyn on every developer's machine, adding to build time overhead.

Who is complaining about this? This is the crux of my confusion here. For developers who build from the command line there is already a solution for exactly this: `-skipAnalyzers`. 

This is frustrating because we have several active threads going on right now about the usability of Visual Studio given the negative performance of our analyzers. Developers who are actively blocked in their work. Yet instead of addressing that (build from VS) we're sending PRs to disable build from command line (which I haven't seen anyone complain about). 

Am I missing the complaints?
Do we have a story for fixing the VS build slow down?

**Comment by mavasani** [comment]
> Do we have a story for fixing the VS build slow down?

I think the confusion is in the wordings that I used in this PR. Sorry, let me try to clarify:

1. "Command line build": I meant an explicit built, both those done inside VS and from a command line prompt - basically, all execution of analyzers from csc.exe/vbc.exe.
2. "Design time build": This is the background build that happens in VS while you are typing, and analyzers need to execute in here to get code style suggestions for active file.

File: eng/targets/Imports.targets [hunk 3]
```diff
@@ -30,6 +30,12 @@
   <PropertyGroup Condition="'$(DisableNullableWarnings)' == 'true'">
     <NoWarn>$(NoWarn);Nullable</NoWarn>
   </PropertyGroup>
+  
+  <!-- Workaround for https://github.com/dotnet/roslyn/issues/42166 -->
+  <!-- PERF: This property group prevents the analyzers from CodeStyle NuGet package from being executed by default in CI -->
```
Comment (mavasani) [inline eng/targets/Imports.targets:35]:
```suggestion
  <!-- PERF: This property group prevents the analyzers from CodeStyle NuGet package from being executed by default in all explicit builds (i.e. csc/vbc invocations both inside Visual Studio and from command line prompt) -->
```

File: eng/targets/Imports.targets [hunk 4]
```diff
@@ -30,6 +30,12 @@
...
...IDE0022;IDE0023;IDE0024;IDE0025;IDE0026;IDE0027;IDE0028;IDE0028WithoutSuggestion;IDE0029;IDE0029WithoutSuggestion;IDE0030;IDE0030WithoutSuggestion;IDE0031;IDE0031WithoutSuggestion;IDE0032;IDE0032WithoutSuggestion;IDE0033;IDE0033WithoutSuggestion;IDE0034;IDE0034WithoutSuggestion;IDE0035;IDE0035WithoutSuggestion;IDE0036;IDE0036WithoutSuggestion;IDE0037;IDE0037WithoutSuggestion;IDE0040;IDE0040WithoutSuggestion;IDE0041;IDE0041WithoutSuggestion;IDE0042;IDE0042WithoutSuggestion;IDE0043;IDE0044;IDE0044WithoutSuggestion;IDE0045;IDE0045WithoutSuggestion;IDE0046;IDE0046WithoutSuggestion;IDE0047;IDE0048;IDE0048WithoutSuggestion;IDE0050;IDE0050WithoutSuggestion;IDE0051;IDE0052;IDE0054;IDE0054WithoutSuggestion;IDE0055;IDE0056;IDE0056WithoutSuggestion;IDE0057;IDE0057WithoutSuggestion;IDE0058;IDE0059;IDE0060;IDE0061;IDE0062;IDE0062WithoutSuggestion;IDE0063;IDE0063WithoutSuggestion;IDE0064;IDE0065;IDE0066;IDE0066WithoutSuggestion;IDE0070;IDE0070WithoutSuggestion;IDE0071;IDE0071WithoutSuggestion;IDE0072;IDE0072WithoutSuggestion;IDE0073;IDE0073WithoutSuggestion;IDE0074;IDE0074WithoutSuggestion;IDE0075;IDE0075WithoutSuggestion;IDE1005;IDE1005WithoutSuggestion;IDE1006;IDE1006WithoutSuggestion</NoWarn>
```
Comment (jaredpar) [inline eng/targets/Imports.targets:37]: How does adding these to `$(NoWarn)` disable the diagnostic?

Comment (mavasani) [reply eng/targets/Imports.targets:37]: If all diagnostics reported by an analyzer are disabled (via nowarn or ruleset) then the analyzer is completely skipped for execution.

Comment (jaredpar) [reply eng/targets/Imports.targets:37]: How does this interact with the editor config change which bumps them all back to warning?

Comment (mavasani) [reply eng/targets/Imports.targets:37]: We currently enforce only 2 of these as warnings in Roslyn's editorconfig (file header and formatting analyzer). Editorconfig based settings always take precedence over compilation options (nowarn and ruleset settings). So those 2 analyzers will run everywhere, and rest will only run while typing in the IDE.

**Comment by mavasani** [comment]
Thanks!

**Pull request merged**